### ISSUE TITLE
[IDEA] Tiddler with same name but different version

### ISSUE THREAD

**Issue by Parlazo** [Jermolene/TiddlyWiki5#5702]
**Is your feature request related to a problem? Please describe.**
The "problem" is...
I make a macro and it is in use
now i had a new idea for the same macro
now i would like to try things and then add the macro but keep the old version of the macro as reference
may later i will keep the new version or i will delete the new version but i want to keep the old version still as reference.
but if two tiddlers have the same name TW complains and want to overwrite the old macro version

**Describe the solution you'd like**
i would like the possibility to create serval tiddlers with the same name
TW should act as normal and warn if a new tiddler has the same name as an older tiddler and also overwrite the older tiddler if the user wants this
but... but if the older tiddler has an other version value in the version field than the newer tiddler it should accept tiddlers with the same name but different version
if two tiddlers have the same name but different version there should appear some where a little selectbox beside the tittle or bottom where all different tiddler versions with the same name are selectable ... if i click a version it should display the specific version in the current edit box or maybe optional in the normal view display too
if such a tiddler is some how active like a macro or template or theme and so on it should be stronger than every older version and TW should use the newest version according to semantic versioning... similar to plugins which overwrite core functions but restore the core functions if the plugin gets deleted
so at the end one could have a macro with the same name but the newest version will be active, while the older versions are viewable via the selectbox and are some how in sleep mode

**Describe alternatives you've considered**
endless titles with versioning, hundreds open tiddlers, endless scrolling, confusion

**Additional context**
such a functionality could also be useful in content like small books
every book page could be a number in the version string but only one tiddler with one name
every page would be a tiddler with the same name gut a different number
the user could read a tiddler or page under one tiddler name and then select a other page under the same name

so a tiddler 

`This is the book you must read`
```
here is a lot of text
and it is the content
of one page
```
`select box ->`
`page 1`
`page 2`
`page 3`

page 1 has a version field with value 1
page 2 has a version field with value 2
page 3 has a version field with value 3

now create a new tiddler "This is the book you must read" TW will create it
now create a new tiddler "This is the book you must read" TW will overwrite the older
now create a new tiddler "This is the book you must read" with version field 1 TW will create it
now create a new tiddler "This is the book you must read" with version field 2 TW will create it
now create a new tiddler "This is the book you must read" TW will overwrite the older one a second time
now create a new tiddler "This is the book you must read" with version field 1 TW will overwrite the old version 1
now create a new tiddler "This is the book you must read" with version field 3 TW will create it

if it is a macro the newest number will be the active and win over older and still existing versions

so if one has a macro bundle he may like delete in fact first all older versions but still keep the title
so he could maybe first install a tiddler "This is the book you must read" with version field * TW will delete all versions and create a empty tiddler
then he can install "This is the book you must read" with no version or with a version as before

**Comment by pmario** [comment]
It is possible to have 2 tiddlers with the same name. We call it a shadow tiddler. Shadow tiddlers come with plugins. If you overwrite them, your changed version wins. 

If you messed up your version, you only need to delete the tiddler and the shadow will take over again. 

That's not 100% what you describe, but close.

**Comment by pmario** [comment]
The easiest way to test a new version of a macro is this:

```
\define test()
some text
\end

\define test()
new text
\end
```

The second version will win. ... BUT that's only a temporary solution. for macro testing.

**Comment by pmario** [comment]
The best way is: Clone your macro tiddler and rename it to `xxx - V0.0.1` and remove the `$:/tags/Macro`, so it will not be active. 

Your new tiddler can be `xxx - V0.1.0` so you know it is the actual version.

**Comment by Jermolene** [comment]
Hi @Parlazo TiddlyWiki uses the "title" field as a primary key; I think you are suggesting changing things so that a combination of a new "version" field and the "title" field is used as a primary key. It's really too late for such a fundamental change to TiddlyWiki 5, but it's interesting to discuss.

As it happens, I do want to add support for retaining a full version history of tiddlers, but my current thinking is along different lines (specifically, for efficiency I'd want to keep the old revisions as a series of diffs, rather than literal text).

**Comment by Parlazo** [comment]
> a combination of a new "version" field and the "title" field is used as a primary key.

Yes exactly ... first the title and (hidden) the version field as one primary key.

> As it happens, I do want to add support for retaining a full version history of tiddlers, but my current thinking is along different lines (specifically, for efficiency I'd want to keep the old revisions as a series of diffs, rather than literal text).

This sounds very interesting too. Maybe there is a way to combine the title+version primary key idea with your series of diffs idea, this sounds very promising 😃 .

**Comment by linonetwo** [comment]
![截屏2023-09-05 17 10 45](https://github.com/Jermolene/TiddlyWiki5/assets/3746270/9811a015-66db-47f2-bfb4-fa15c362be5f)

I recommend using external backup, for example, .git, to do this.

My main wiki is 50M, and the .git backup is 200M. If this happened inside wiki, then my wiki will be 261M.

I always want to write a plugin to show old version in the .git. The drawback is this can only be used in Desktop Nodejs wiki, because git can only be used easily in Desktop environment...

**Comment by pmario** [comment]
See: https://github.com/TiddlyWiki/TiddlyWiki5/issues/5702#issuecomment-843965596
